For collect2, take -G and -bsvr4 to change library-extension search order to
what -brtl does. Also, take -bnortl to disable -brtl mode. Upstream report is
https://gcc.gnu.org/ml/gcc-patches/2013-02/msg00104.html

For collect2, ignore shared archive members with the F_LOADONLY flag.

diff -ruN gcc-4.2.4/gcc/collect2.c /junk/base-2010.0u/portage/sys-devel/gcc-4.2.4-r01.11/work/gcc-4.2.4/gcc/collect2.c
--- gcc-4.2.4/gcc/collect2.c	2007-12-31 17:04:26 +0100
+++ /junk/base-2010.0u/portage/sys-devel/gcc-4.2.4-r01.11/work/gcc-4.2.4/gcc/collect2.c	2014-06-30 15:51:22 +0200
@@ -1068,8 +1068,15 @@
 		export_flag = 1;
 	      else if (arg[2] == '6' && arg[3] == '4')
 		aix64_flag = 1;
-	      else if (arg[2] == 'r' && arg[3] == 't' && arg[4] == 'l')
-		aixrtl_flag = 1;
+	      else if (arg[2] == 'r' && arg[3] == 't' && arg[4] == 'l' && arg[5] == '\0')
+		aixrtl_flag = 1; /* -brtl enables (but not -brtllib) */
+	      else if (arg[2] == 'n' && arg[3] == 'o' && arg[4] == 'r' && arg[5] == 't' && arg[6] == 'l' && arg[7] == '\0')
+		aixrtl_flag = 0; /* -bnortl disables (but not -bnortllib) */
+	      else if (arg[2] == 's' && arg[3] == 'v' && arg[4] == 'r' && arg[5] == '4')
+		aixrtl_flag = 1; /* -bsvr4 enables -brtl too */
+	      break;
+	    case 'G':
+	      aixrtl_flag = 1; /* -G enables -brtl too */
 	      break;
 #endif
 
@@ -2322,12 +2329,14 @@
 /* 0757 = U803XTOCMAGIC (AIX 4.3) and 0767 = U64_TOCMAGIC (AIX V5) */
 #ifdef _AIX51
 #   define GCC_CHECK_HDR(X) \
-     ((HEADER (X).f_magic == U802TOCMAGIC && ! aix64_flag) \
-      || (HEADER (X).f_magic == 0767 && aix64_flag))
+     (((HEADER (X).f_magic == U802TOCMAGIC && ! aix64_flag) \
+       || (HEADER (X).f_magic == 0767 && aix64_flag)) \
+      && !(HEADER (X).f_flags & F_LOADONLY))
 #else
 #   define GCC_CHECK_HDR(X) \
-     ((HEADER (X).f_magic == U802TOCMAGIC && ! aix64_flag) \
-      || (HEADER (X).f_magic == 0757 && aix64_flag))
+     (((HEADER (X).f_magic == U802TOCMAGIC && ! aix64_flag) \
+       || (HEADER (X).f_magic == 0757 && aix64_flag)) \
+      && !(HEADER (X).f_flags & F_LOADONLY))
 #endif
 
 #endif
